src/commands/repos/UpdateRepos.ts

```diff
@@ -126,42 +127,59 @@ export class UpdateRepos extends AbstractReposCommand {
             await repo.fetch({branch: repoProperties.branch});
             await repo.checkoutCommit(repoProperties.commit);
-        } else if (repoProperties.tag || repoProperties.latestTagForRelease) {
-            const tagToCheckout = repoProperties.tag || repoProperties.latestTagForRelease;
-            await repo.resetHard();
-            await repo.fetch({tag: tagToCheckout});
-            await repo.checkoutTag(tagToCheckout);
-            await repo.createBranchForTag(tagToCheckout);
+        } else if (repoProperties.tag) {
+            console.log(`[${repoName}]: will update to the tag ${repoProperties.tag}.`);
+            await this.checkoutTag(repo, repoName, repoProperties.tag);
+        } else if (repoProperties.latestTagForRelease) {
+            console.log(`[${repoName}]: will update to the latestTagForRelease ${repoProperties.latestTagForRelease}.`);
+            Repository.validateTagMarker(repoProperties, repoName);
+            await this.checkoutTag(repo, repoName, repoProperties.latestTagForRelease);
         } else {
-            // checkout branch
-            await repo.resetHard();
-            await repo.checkoutBranch(repoProperties.branch);
-            if (this.resetToRemote) {
-                await repo.resetHard(repoProperties.branch);
-            } else if (this.noFetch) {
-                // don't update to the remote branch, but stay at the current local branch
-            } else {
-                await repo.pullOnlyFastForward(repoProperties.branch);
-            }
+            console.log(`[${repoName}]: will update to the latest HEAD of remote branch ${repoProperties.branch} because no latestTagForRelease was found and only a branch is configured.`);
+            await this.checkoutBranch(repo, repoProperties);
+        }
+    }
+
+    private async checkoutTag(repo: Repository, repoName: string, tagToCheckout: string): Promise<void> {
+        await repo.resetHard();
+        await repo.fetch({tag: tagToCheckout});
+        await repo.checkoutTag(tagToCheckout);
+        await repo.createBranchForTag(repoName, tagToCheckout);
+    }
+
+    private async checkoutBranch(repo: Repository, repoProperties: IRepoStatus): Promise<void> {
+        // checkout branch
+        await repo.resetHard();
+        await repo.prefetchBranchForShallowClone(repoProperties.branch);
+        await repo.checkoutBranch(repoProperties.branch);
+        if (this.resetToRemote) {
+            await repo.resetHard(repoProperties.branch);
+        } else if (this.noFetch) {
+            // don't update to the remote branch, but stay at the current local branch
+        } else {
+            await repo.pullOnlyFastForward(repoProperties.branch);
         }
     }
 
-    private async getTargetBranch(repo: Repository, repoProperties: IRepoStatus): Promise<string> {
-        let targetBranch: string;
+    private async getTargetRef(repo: Repository, repoProperties: IRepoStatus): Promise<string> {
```

Comment:
This method now seems like duplication of the logic in the `checkout` method. Also the logic is not quite the same so it could lead to a wrong outcome.

Since this method is used only to determine if the `node_modules` should be deleted, maybe it can be removed.  Checking the node_modules can happen right before doing the actual checkout in the `checkout` method (once it's determined exactly what should be checked out).

Reply:
actually I just renamed it. currently I would prefer to not change the old behaviour to avoid a feature freeze. but yes, this could be optimised.

Reply:
hm... I'll check how complicated that change would be

Reply:
As tests now fail I rolled back that change. We should keep that in mind for optimisation later.

Reply:
See [PFM-ISSUE-11753](https://base.cplace.io/pages/v29mvuvno0vvlmyqj8qfpvloz/PFM-ISSUE-11753-cplace-cli-optimize-node-module-cleanup?highlightedEntityUid=page%2Fv29mvuvno0vvlmyqj8qfpvloz)